File: frontend/src/types/endpoint.ts
/**
 * TypeScript types for learnable workflow endpoints.
 */

export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'DELETE';
export type TransformMode = 'direct' | 'code';

/**
 * Full endpoint model returned from API.
 */
export interface Endpoint {
  id: string;
  workflowId: string;
  name: string;
  slug: string;
  description?: string;
  httpMethod: HttpMethod;
  instruction: string;
  mode: TransformMode;

  // Learning status
  isLearned: boolean;
  learnedAt?: string;
  learnedSkillMd?: string; // Only included on detail requests
  learnedTransformerCode?: string; // Only for code mode, detail requests

  // Metadata
  createdAt: string;
  updatedAt: string;
  lastExecutedAt?: string;
  executionCount: number;
}

/**
 * Request to create a new endpoint.
 */
export interface EndpointCreate {
  name: string;
  slug: string;
  description?: string;
  httpMethod?: HttpMethod;
  instruction: string;
  mode?: TransformMode;
}

/**
 * Request to update an endpoint.
 */
export interface EndpointUpdate {
  name?: string;
  description?: string;
  httpMethod?: HttpMethod;
  instruction?: string;
  mode?: TransformMode;
}

/**
 * Response for listing endpoints.
 */
export interface EndpointsResponse {
  endpoints: Endpoint[];
  total: number;
}

/**
 * Request body for executing an endpoint.
 */
export interface EndpointExecuteRequest {
  inputData?: Record<string, unknown> | unknown[] | string;
  learn?: boolean;
  apply?: boolean;
}

/**
 * Response from endpoint execution.
 */
export interface EndpointExecuteResponse {
  success: boolean;
  result?: Record<string, unknown> | unknown[];
  nodesCreated: number;
  nodesUpdated: number;
  nodesDeleted: number;
  edgesCreated: number;
  errors: string[];
  executionTimeMs: number;
}

/**
 * SSE event from endpoint execution stream.
 */
export interface EndpointExecutionEvent {
  event: string;
  phase?: string;
  message?: string;
  tool?: string;
  input?: Record<string, unknown>;
  result?: string;
  valid?: boolean;
  itemCount?: number;
  errors?: string[];
  skillMd?: string;
  nodesCreated?: number;
  nodesUpdated?: number;
  nodesDeleted?: number;
  edgesCreated?: number;
  executionTimeMs?: number;
}

/**
 * Saved request preset for API tester.
 */
export interface RequestPreset {
  id: string;
  name: string;
  endpointId: string;
  headers: Record<string, string>;
  params: Record<string, string>;
  body: string;
  createdAt: string;
}

/**
 * Request history entry for API tester.
 */
export interface RequestHistoryEntry {
  id: string;
  endpointId: string;
  endpointSlug: string;
  httpMethod: HttpMethod;
  url: string;
  headers: Record<string, string>;
  body: string;
  response: {
    status: number;
    body: string;
    timeMs: number;
  };
  timestamp: string;
}

/**
 * Seed node from transformer output (for POST endpoints).
 */
export interface SeedNode {
  temp_id: string;
  node_type: string;
  title: string;
  status?: string;
  properties: Record<string, unknown>;
}

/**
 * Seed edge from transformer output (for POST endpoints).
 */
export interface SeedEdge {
  edge_type: string;
  from_temp_id: string;
  to_temp_id: string;
  properties?: Record<string, unknown>;
}

/**
 * SeedData result from POST endpoint preview.
 */
export interface SeedData {
  nodes: SeedNode[];
  edges: SeedEdge[];
}

/**
 * Query result from GET endpoint.
 */
export interface QueryResult {
  nodes: Record<string, unknown>[];
  count: number;
}

/**
 * Update result from PUT endpoint.
 */
export interface UpdateResult {
  updates: Array<{
    node_id: string;
    properties: Record<string, unknown>;
  }>;
}

/**
 * Delete result from DELETE endpoint.
 */
export interface DeleteResult {
  node_ids: string[];
}

/**
 * Pending result from preview mode execution.
 */
export interface PendingResult {
  transformResult: SeedData | QueryResult | UpdateResult | DeleteResult;
  transformCode?: string;
  skillMd?: string;
  httpMethod: HttpMethod;
  // Preview counts
  nodesCreated?: number;
  edgesCreated?: number;
  nodesUpdated?: number;
  nodesDeleted?: number;
  nodesToCreate?: SeedNode[];
  edgesToCreate?: SeedEdge[];
  updatesToApply?: Array<{ node_id: string; properties: Record<string, unknown> }>;
  nodesToDelete?: string[];
}

/**
 * Request body for applying a previewed result.
 */
export interface ApplyPreviewRequest {
  transformResult: Record<string, unknown>;
}

/**
 * Response from applying a previewed result.
 */
export interface ApplyPreviewResponse {
  success: boolean;
  nodesCreated: number;
  nodesUpdated: number;
  nodesDeleted: number;
  edgesCreated: number;
  errors: string[];
}

File: backend/app/models/endpoint.py
"""Pydantic models for learnable workflow endpoints."""

from typing import Any, Literal

from pydantic import BaseModel, Field


class EndpointCreate(BaseModel):
    """Request to create a new endpoint."""

    model_config = {"populate_by_name": True}

    name: str = Field(..., description="Human-readable name for the endpoint")
    slug: str = Field(..., pattern=r"^[a-z0-9-]+$", description="URL-safe identifier")
    description: str | None = Field(None, description="What this endpoint does")
    http_method: Literal["GET", "POST", "PUT", "DELETE"] = Field(
        "POST", alias="httpMethod", description="HTTP method for this endpoint"
    )
    instruction: str = Field(..., description="Natural language transformer instruction")
    mode: Literal["direct", "code"] = Field(
        "direct", description="Transformer mode: direct or code"
    )


class EndpointUpdate(BaseModel):
    """Request to update an endpoint."""

    model_config = {"populate_by_name": True}

    name: str | None = None
    description: str | None = None
    http_method: Literal["GET", "POST", "PUT", "DELETE"] | None = Field(
        None, alias="httpMethod"
    )
    instruction: str | None = None
    mode: Literal["direct", "code"] | None = None


class Endpoint(BaseModel):
    """Full endpoint model."""

    model_config = {"populate_by_name": True, "serialize_by_alias": True}

    id: str
    workflow_id: str = Field(alias="workflowId")
    name: str
    slug: str
    description: str | None = None
    http_method: Literal["GET", "POST", "PUT", "DELETE"] = Field(alias="httpMethod")
    instruction: str
    mode: Literal["direct", "code"]

    # Learning status
    is_learned: bool = Field(
        ..., alias="isLearned", description="Whether the endpoint has been learned"
    )
    learned_at: str | None = Field(None, alias="learnedAt")
    learned_skill_md: str | None = Field(
        None,
        alias="learnedSkillMd",
        description="Generated SKILL.md content (only included on detail requests)",
    )
    learned_transformer_code: str | None = Field(
        None,
        alias="learnedTransformerCode",
        description="Generated transform.py code (only for code mode, detail requests)",
    )

    # Metadata
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")
    last_executed_at: str | None = Field(None, alias="lastExecutedAt")
    execution_count: int = Field(0, alias="executionCount")


class EndpointExecuteRequest(BaseModel):
    """Request body for executing an endpoint."""

    model_config = {"populate_by_name": True}

    input_data: dict[str, Any] | list[Any] | str | None = Field(
        None, alias="inputData", description="Input data for the transformation"
    )
    learn: bool = Field(False, description="If true, run full transformer and save skill")
    apply: bool = Field(
        True, description="If false, return preview without applying changes"
    )


class EndpointExecuteResponse(BaseModel):
    """Response from endpoint execution."""

    model_config = {"populate_by_name": True, "serialize_by_alias": True}

    success: bool
    result: dict[str, Any] | list[Any] | None = Field(
        None, description="Query results (for GET endpoints)"
    )
    nodes_created: int = Field(0, alias="nodesCreated")
    nodes_updated: int = Field(0, alias="nodesUpdated")
    nodes_deleted: int = Field(0, alias="nodesDeleted")
    edges_created: int = Field(0, alias="edgesCreated")
    errors: list[str] = Field(default_factory=list)
    execution_time_ms: int = Field(0, alias="executionTimeMs")


class EndpointsResponse(BaseModel):
    """Response for listing endpoints."""

    model_config = {"serialize_by_alias": True}

    endpoints: list[Endpoint]
    total: int


class ApplyPreviewRequest(BaseModel):
    """Request body for applying a previewed endpoint result."""

    model_config = {"populate_by_name": True}

    transform_result: dict[str, Any] = Field(
        ..., alias="transformResult", description="The transform result from preview"
    )
    match_result: dict[str, Any] | None = Field(
        None, alias="matchResult", description="Optional matching results for POST endpoints"
    )


class ApplyPreviewResponse(BaseModel):
    """Response from applying a previewed endpoint result."""

    model_config = {"populate_by_name": True, "serialize_by_alias": True}

    success: bool
    nodes_created: int = Field(0, alias="nodesCreated")
    nodes_updated: int = Field(0, alias="nodesUpdated")
    nodes_deleted: int = Field(0, alias="nodesDeleted")
    edges_created: int = Field(0, alias="edgesCreated")
    errors: list[str] = Field(default_factory=list)
